openproblems/tasks/chromatin_potential/methods/linear_regression.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import pandas as pd
+from scipy.sparse import csr_matrix
+import scanpy as sc
+import subprocess
+from ....tools.decorators import method
+
+
+def _chrom_limit(x, tss_size=2e5):
+    """Extending TSS to upstream and downstream intervals
+
+    Parameters
+    ----------
+    x : pd.Series
+        a pd.Series containing [start, end, direction] where start and end are ints and direction is {'+', '-'}.
+    tss_size: int
+        a int that defines the upstream and downstream extending regions around TSS
+    """
+    y = x.values
+    gene_direction = y[-1]
+    gene_start = y[-3]
+    gene_end = y[-2]
+    if gene_direction == "+":
+        return [gene_start - tss_size // 2, gene_start + tss_size // 2]
+    else:
+        return [gene_end - tss_size // 2, gene_end + tss_size // 2]
+
+
+def _get_annotation(adata):
+    """This function insert meta data into .obs for adata"""
+    from pyensembl import EnsemblRelease
+
+    # TODO: add more species
+    subprocess.call(
+        [
+            "pyensembl",
+            "install",
+            "--release",
+            adata.uns["release"],
+            "--species",
+            adata.uns["species"],
+        ]
+    )
+    data = EnsemblRelease(adata.uns["release"], species=adata.uns["species"])
+
+    # get ensemble gene coordinate
+    genes = []
+    for i in adata.var.index.map(lambda x: x.split(".")[0]):
+        try:
+            gene = data.gene_by_id(i)
+            genes.append(
+                [
+                    "chr%s" % gene.contig,
+                    gene.start,
+                    gene.end,
+                    gene.strand,
+                ]
+            )
+        except:
+            genes.append([np.nan, np.nan, np.nan, np.nan])
+    old_col = adata.var.columns.values
+    adata.var = pd.concat(
+        [adata.var, pd.DataFrame(genes, index=adata.var_names)], axis=1
+    )
+    adata.var.columns = np.hstack(
+        [old_col, np.array(["chr", "start", "end", "strand"])]
+    )
+
+
+def _atac_genes_score(adata, top_genes=500, threshold=1):
+    """This function calculate gene score and insert into .obsm"""
+    import pybedtools
+
+    # get annotation for TSS
+    _get_annotation(adata)
+    # basic quality control
+    sc.pp.filter_cells(adata, min_genes=200)
+    sc.pp.filter_genes(adata, min_cells=5)
+
+    adata.var["mt"] = adata.var.gene_short_name.str.startswith(
+        "mt-"
+    )  # annotate the group of mitochondrial genes as 'mt'
```

Comment:
For species compatibility can we accept "MT-" too? Maybe with `adata.var.gene_short_name.str.lower().str.startswith`

Reply:
Great suggestion, will modify this soon